Stop doing sync over async in Produce100Continue

wtgodbe opened the pull request:
Fixes https://github.com/dotnet/aspnetcore/issues/31225

Draft with a first-pass at fixing the issue.

wtgodbe committed:
Stop doing sync over async in Produce100Continue

wtgodbe committed:
Address feedback

wtgodbe committed:
IsCompleted -> IsCompletedSuccessfully

wtgodbe committed:
Remove unneeded awaited methods

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs:
```diff
@@ -132,7 +132,7 @@ public override bool TryReadInternal(out ReadResult readResult)
             }
 
-            TryStart();
+            TryStartAsync();
```
Isn't this a behavior change to go from blocking to fire-and-forget?  Can TryStartAsync throw asynchronously like if the client has disconnected?

halter73 replied:
If `Write100ContinueAsync()` could realistically return an incomplete ValueTask this is a change in behavior. Not sure how that could ever happen though. If it could happen, not blocking is waaay better than block because at least you return the thread to the pool. TryStartAsync should never be able to throw asynchronously because it's a write without a canceled token.

davidfowl replied:
Add an assert. I dislike the fire and forget as well unless we have an explicit check that it's always completed.

wtgodbe replied:
https://github.com/dotnet/aspnetcore/pull/31650/commits/70e8ee1a5fe32cbb628bd809e0f74c55fefdf6f3

wtgodbe replied:
Whoops, plus https://github.com/dotnet/aspnetcore/pull/31650/commits/a10c9d999e6d687d85c3d057f6fee35e8fef03c1

davidfowl replied:
I love that bug 😄

halter73 replied:
> I dislike the fire and forget

I totally agree. I also dislike fire-and-forget with a passion, but what's even worse is blocking on I/O. This isn't the [only](https://github.com/dotnet/aspnetcore/blob/1d7443d1823a63b60f94ec9e93b1c4ca8e635aae/src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/StreamInputFlowControl.cs#L60) [place](https://github.com/stephentoub/aspnetcore/blob/b5145963a8a44832bf2d49a0beb1e6de0acb4cd5/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs#L147-L148) we don't await  output-writing `ValueTask`s we expect to be usually completed.

Now unlike with HTTP/2 where you could conceive of a scenario where output backpressure has built up prior to sending a `WINDOW_UPDATE` or `RST_STREAM`, it's hard to come up with one for an HTTP/1.1 "100 continue" response. That said, it's not a mathematical invariant. I could write a unit test with a custom transport and write buffering disabled to force it. For that reason, we should have this Debug.Assert like we don't for `WINDOW_UPDATE` or `RST_STREAM`. If we think it's worthwhile we can store the task and await it in ConsumeAsync(), but I don't think that's necessary.

wtgodbe replied:
That makes sense to me - if it matches what we do for `RST_STREAM` and `WINDOW_UPDATE` seems like we can remove the assert here

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http/Http1ChunkedEncodingMessageBody.cs:
```diff
@@ -101,7 +101,8 @@ private async Task PumpAsync()
                 if (!awaitable.IsCompleted)
                 {
-                    TryProduceContinue();
+                    ValueTask<FlushResult> continueTask = TryProduceContinueAsync();
+                    await continueTask;
```
```suggestion
                    await TryProduceContinueAsync();
```

wtgodbe replied:
https://github.com/dotnet/aspnetcore/pull/31650/commits/70e8ee1a5fe32cbb628bd809e0f74c55fefdf6f3

wtgodbe committed:
Feedback

wtgodbe committed:
Brain fart

wtgodbe commented:
Got what looks to be a flaky failure in a selenium test, `InputDateInteractsWithEditContext_NonNullableDateTime`:

> OpenQA.Selenium.BrowserAssertFailedException : Xunit.Sdk.EqualException: Assert.Equal() Failure\r\nExpected: modified valid\r\nActual:   modified invalid\r\n   at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer1 comparer) in C:\\Dev\\xunit\\xunit\\src\\xunit.assert\\Asserts\\EqualityAsserts.cs:line 40\r\n at Microsoft.AspNetCore.E2ETesting.WaitAssert.<>c__DisplayClass17_0.<WaitAssertCore>b__0() in /_/src/Shared/E2ETesting/WaitAssert.cs:line 83\r\n at Microsoft.AspNetCore.E2ETesting.WaitAssert.<>c__DisplayClass18_01.<WaitAssertCore>b__0(IWebDriver ) in //src/Shared/E2ETesting/WaitAssert.cs:line 109\r\nScreen shot captured at 'D:\workspace\_work\1\s\artifacts\TestResults\Release\Microsoft.AspNetCore.Components.E2ETests\bf12e230a077482c93a44465ccda36f2.png'\r\nEncountered browser errors\r\n[2021-04-12T17:01:25Z] [Info] http://127.0.0.1:53621/subdir/_framework/blazor.server.js 0:31687 "[2021-04-12T17:01:25.992Z] Information: Normalizing '_blazor' to 'http://127.0.0.1:53621/subdir/_blazor'."\r\n[2021-04-12T17:01:25Z] [Info] http://127.0.0.1:53621/subdir 45:16 "Blazor server-side"\r\n[2021-04-12T17:01:26Z] [Info] http://127.0.0.1:53621/subdir/_framework/blazor.server.js 0:31687 "[2021-04-12T17:01:25.998Z] Information: WebSocket connected to ws://127.0.0.1:53621/subdir/_blazor?id=foJb3ozHB7bXxtdKAJ8Kbg."\r\n[2021-04-12T17:01:27Z] [Info] http://127.0.0.1:53621/subdir/_framework/blazor.server.js 0:31687 "[2021-04-12T17:01:27.398Z] Information: Normalizing '_blazor' to 'http://127.0.0.1:53621/subdir/_blazor'."\r\n[2021-04-12T17:01:27Z] [Info] http://127.0.0.1:53621/subdir 45:16 "Blazor server-side"\r\n[2021-04-12T17:01:27Z] [Info] http://127.0.0.1:53621/subdir/_framework/blazor.server.js 0:31687 "[2021-04-12T17:01:27.408Z] Information: WebSocket connected to ws://127.0.0.1:53621/subdir/_blazor?id=n_74YY7aNbRQ53NAUSdGuQ."Page content:\r\n<head>\r\n    <meta charset="utf-8">\r\n    <title>Basic test app</title>\r\n    <base href="/subdir/">\r\n    <link href="style.css" rel="stylesheet">\r\n    <link rel="icon" href="data:,">\r\n\r\n    \r\n    <link href="_content/TestContentPackage/styles.css" rel="stylesheet">\r\n    <link href="Components.TestServer.styles.css" rel="stylesheet">\r\n</head>\r\n<body>\r\n    <root><div id="test-selector">\r\n    Select test:\r\n    <select id="test-selector-select"><option value="none">Choose...\r\n        <option value="BasicTestApp.AfterRenderInteropComponent">After-render interop component\r\n        <option value="BasicTestApp.AuthTest.AuthRouter">Auth cases\r\n        <option value="BasicTestApp.BindCasesComponent">bind cases\r\n        <option value="BasicTestApp.ComponentRefComponent">Component ref component\r\n        <option value="BasicTestApp.ConfigurationComponent">Con

CC @captainsafia since you're on build ops

Other than that, any other concerns @Tratcher @davidfowl?

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http/Http1ChunkedEncodingMessageBody.cs:
```diff
@@ -44,7 +44,7 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
         public override bool TryReadInternal(out ReadResult readResult)
         {
-            TryStart();
+            TryStartAsync();
```
Assert IsCompleted here too.

wtgodbe replied:
Done

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs:
```diff
@@ -59,7 +65,7 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
         public override bool TryRead(out ReadResult readResult)
         {
-            TryStart();
+            TryStartAsync();
```
Do we expect HTTP/2&3 will also complete this synchronously?

jkotalik replied:
Either way, add the asserts to verify?

wtgodbe replied:
I'm not aware of a reason why we wouldn't expect that - I can add the assert here also

davidfowl replied:
Add the assert yeah

wtgodbe replied:
Done

wtgodbe committed:
Update Http1ChunkedEncodingMessageBody.cs

wtgodbe committed:
Update Http2MessageBody.cs

wtgodbe committed:
Update Http2MessageBody.cs

wtgodbe committed:
Update Http1ContentLengthMessageBody.cs

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs:
```diff
@@ -128,7 +140,13 @@ protected void TryStart()
             }
 
-            OnReadStarted();
+            Task readTask = OnReadStartedAsync();
+            if (!readTask.IsCompletedSuccessfully)
+            {
+                return readTask;
+            }
+
+            return Task.CompletedTask;
```
Why not:
```suggestion
            return OnReadStartedAsync();
```
If the OnReadStartedAsync task is complete then return just return it. And if it is not complete then also just return it...

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs:
```diff
@@ -183,7 +202,11 @@ protected ValueTask<ReadResult> StartTimingReadAsync(ValueTask<ReadResult> readA
             if (!readAwaitable.IsCompleted)
             {
-                TryProduceContinue();
+                ValueTask<FlushResult> continueTask = TryProduceContinueAsync();
```
Shouldn't we indicate that the ValueTask result has been consumed? @stephentoub submitted a PR to do that for historical code: https://github.com/dotnet/aspnetcore/pull/31221

```cs
if (!continueTask.IsCompletedSuccessfully)
{
    return StartTimingReadAwaited(continueTask, readAwaitable, cancellationToken);
}
else
{
    continueTask.GetAwaiter().GetResult();
}
```

stephentoub replied:
Hmm.  I'd have expected the code in the PR to trigger CA2012, providing the same feedback as James.  Were no warnings issued for this?

JamesNK replied:
Your .editconfig changes are still there: https://github.com/dotnet/aspnetcore/blob/main/.editorconfig

I'm not familiar with configuring analyzers in editconfigs or CA2021. I'm not sure why it wasn't prompted.

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs:
```diff
@@ -30,13 +31,19 @@ protected override void OnReadStarting()
             if (!_context.RequestBodyStarted)
             {
-                TryProduceContinue();
+                ValueTask<FlushResult> continueTask = TryProduceContinueAsync();
+                if (!continueTask.IsCompletedSuccessfully)
+                {
+                    return continueTask.AsTask();
+                }
```
We have a helper method for `ValueTask<FlushResult>` in this situation. Something like `continueTask.GetAsTask()`

JamesNK replied:
https://github.com/dotnet/aspnetcore/blob/d94cee250d7df317fa5b87b1a2396ff9dbb08283/src/Shared/ValueTaskExtensions/ValueTaskExtensions.cs

JamesNK replied:
```suggestion
                return TryProduceContinueAsync().GetAsTask();
```

wtgodbe committed:
Respond to feedback

wtgodbe commented:
@JamesNK @halter73 just pushed https://github.com/dotnet/aspnetcore/pull/31650/commits/a9a561943dec871701426b40b572123d06a3cfdb to respond to feedback

wtgodbe commented:
Got a real test failure in `Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http3StreamTests.ContentLength_Received_NoDataFrames_Reset`

> Assert.Equal() Failure
Expected: ProtocolError
Actual:   0

Looking

jkotalik commented:
I think it isn't you: https://github.com/dotnet/aspnetcore/pull/31743#issuecomment-818811824

jkotalik commented:
@JamesNK should we just disable in this PR?

JamesNK commented:
Issue: https://github.com/dotnet/aspnetcore/issues/31777

Add skip or quarantine back in this PR.

wtgodbe committed:
Merge remote-tracking branch 'origin/main' into wtgodbe/100Continue

wtgodbe committed:
Quarantine failling test

wtgodbe commented:
Done

wtgodbe commented:
@JamesNK any other concerns?
